DDF/PE.hs

```diff
@@ -0,0 +1,144 @@
+data P repr h a where
+  Unk :: repr h a -> P repr h a
+  Static :: a -> (forall h. repr h a) -> P repr h a
+  StaFun :: (forall hout. EnvT repr (a, h) hout -> P repr hout b) ->
+            P repr h (a -> b)
+  Open   :: (forall hout. EnvT repr h hout -> P repr hout a) ->
+            P repr h a
+
+data EnvT repr hin hout where
+  Dyn  :: EnvT repr hin hin
+  Arg  :: P repr hout a -> EnvT repr (a, hout) hout
+  Weak :: EnvT repr h (a, h)
+  Next :: EnvT repr hin hout -> EnvT repr (a, hin) (a, hout)
+
+dynamic:: DBI repr => P repr h a -> repr h a
+dynamic (Unk x)      = x
+dynamic (Static _ x) = x
+dynamic (StaFun f)   = abs $ dynamic (f Dyn)
+dynamic (Open f)     = dynamic (f Dyn)
+
+app_open :: DBI repr =>
+            P repr hin r -> EnvT repr hin hout -> P repr hout r
+app_open e Dyn            = Unk (dynamic e)
+            -- this was a constant function, ignore the argument
+app_open (Static es ed) _ = Static es ed
+app_open (Open fs) h      = fs h
+app_open (StaFun fs) h    = abs (fs (Next h))
+app_open (Unk e) h        = Unk (app_unk e h) where
+  -- The reason we need the following
+  --   consider the term (lam (z + dynamic z)) (int 1)
+  -- The PE result should really be 1 + (lam z) (int 1)
+  app_unk :: DBI repr =>
+             repr hin a -> EnvT repr hin hout -> repr hout a
+  app_unk e Dyn      = e
+  app_unk e (Arg p)  = app (abs e) (dynamic p)
+  app_unk e (Next h) = app (s (app_unk (abs e) h)) z
+  app_unk e Weak     = s e
+
+instance DBI r => DBI (P r) where
+  z = Open f where
+    f :: EnvT r (a,h) hout -> P r hout a
+    f Dyn       = Unk z                 -- turn to dynamic as requested
+    f (Arg x)   = x                     -- substitution
+    f (Next _)  = z                     -- not my level
+    f Weak      = s z
+
+  s :: forall h a any. P r h a -> P r (any, h) a
+  s (Unk x) = Unk (s x)
+  s (Static a ar) = Static a ar
+  s (StaFun fs)   = abs (fs (Next Weak))
+  s p = Open f where
+    f :: EnvT r (any, h) hout -> P r hout a
+    -- Nothing is statically known, dynamize
+    f Dyn              = Unk (s (dynamic p))
+    f (Arg _)          = p
+    f (Next h)         = s (app_open p h)
+    f Weak             = s (s p)
+
+  abs (Unk f) = Unk (abs f)
+  abs (Static k ks) = StaFun $ \_ -> Static k ks
+  abs body = StaFun (\args -> app_open body args)
+  
+  app (Unk f) (Unk x) = Unk (app f x)
+  app (StaFun fs) p   = fs (Arg p)
+  app (Static _ fs) p = Unk (app fs (dynamic p))
+  app e1 e2           = Open (\h -> app (app_open e1 h) (app_open e2 h))
+
+-- Try to PE D.Double ...
+
+instance Double r => Double (P r) where
+  double x | trace (M.unwords ["PE: D.double", show x]) M.True =
+               Static x (double x)
+  doublePlus = StaFun (\(Arg a) -> StaFun (\(Arg b) -> f a b))
+    where
+      f (Static d1 _) (Static d2 _) = double (d1 + d2)
+
+pe :: Double repr => P repr () a -> repr () a
+pe = dynamic
+
+t1 :: Double r => r () M.Double
+t1 = app2 doublePlus (double 1.0) (double 2.0)
+-- > runShow t1 [] 1
```

Comment:
Well, the fact is that I don't even know what does the second and third argument mean! Just dropping in random stuff to test quickly. I will refine it a bit later :)